File: prep/implementation/practice_iterations/merge_intervals_2/interval_intersection.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

bool has_overlap(const Interval &a, const Interval &b) {
  return (a.start >= b.start && a.start <= b.end) ||
         (b.start >= a.start && b.start <= a.end);
}

// O(n + m) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, m;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    int s, e;
    for (int i = 0; i < n; i++) {
      cin >> s >> e;
      a.push_back(Interval(s, e));
    }

    cin >> m;
    vector<Interval> b;
    for (int i = 0; i < m; i++) {
      cin >> s >> e;
      b.push_back(Interval(s, e));
    }

    vector<Interval> res;

    int i = 0, j = 0;
    while (i < n && j < m) {
      if (has_overlap(a[i], b[j])) {
        int start = max(a[i].start, b[j].start);
        int end = min(a[i].end, b[j].end);
        res.push_back(Interval(start, end));
      }

      if (a[i].end < b[j].end) {
        i++;
      } else {
        j++;
      }
    }

    for (auto inter : res) {
      cout << inter.start << " " << inter.end << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/merge_intervals_5/IntervalIntersection.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n + m) / O(1)
public class IntervalIntersection {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        List<Interval> a = new ArrayList<>();
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        int m = in.nextInt();
        List<Interval> b = new ArrayList<>();
        for (int i = 0; i < m; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          b.add(new Interval(start, end));
        }

        int i = 0, j = 0;
        List<Interval> res = new ArrayList<>();

        while (i < n && j < m) {
          if ((a.get(i).start >= b.get(j).start &&
               a.get(i).start <= b.get(j).end) ||
              (b.get(j).start >= a.get(i).start &&
               b.get(j).start <= a.get(i).end)) {
            res.add(new Interval(Math.max(a.get(i).start, b.get(j).start),
                                 Math.min(a.get(i).end, b.get(j).end)));
          }

          if (a.get(i).end <= b.get(j).end) {
            i++;
          } else {
            j++;
          }
        }

        for (Interval inter : res) {
          System.out.printf("%d %d\n", inter.start, inter.end);
        }
      }
    }
  }
}